nomt/src/bitbox/wal/mod.rs

```diff
@@ -1,243 +1,132 @@
-
-pub use self::{batch::Batch, entry::Entry};
-
-use self::record::Record;
-
-// WAL format:
-//
-// series of 256KB records
-// each record contains a sequence number, a checksum, a number of entries, and the entries
-// themselves. entries never span records.
-//
-// Record format:
-//   sequence number (8 bytes)
-//   checksum (4 bytes) (magic number ++ last ++ data)
-//   entry count (2 bytes)
-//   LAST (1 byte) | 0xFF if the last record for the sequence, 0x00 otherwise.
-//   data: [Entry]
-//
-// Entry format:
-// kind: u8
-//   1 => updated bucket
-//   2 => cleared bucket
-// data: (variable based on kind)
-//  bucket update:
-//    page ID (16 bytes)
-//    diff (16 bytes)
-//    changed [[u8; 32]] (len with diff)
-//    bucket index (8 bytes)
-//  bucket cleared:
-//    bucket index (8 bytes)
-//
-// Multiple records are expected to have the same sequence number if they are part of the same
-// transaction.
-// e.g. [0 0 0 1 1 2 2 2 2 3 3 4 4 4 4] might be the order of sequence numbers in a WAL.
-// this can be used to compare against the sequence number in the database's metadata file to
-// determine which WAL entries need to be reapplied.
-//
-// WAL entries imply writing buckets and the meta-map. when recovering from a crash, do both.
-
-const WAL_RECORD_SIZE: usize = 256 * 1024;
-// walawala
-const WAL_CHECKSUM_MAGIC: u32 = 0x00a10a1a;
-
-// Open a WAL file and append encoded batch of writes to the end,
-// and prune previous ones if the current succeeded.
-//
-// It does not perform any check on the state of the WAL file
-pub struct WalWriter {
-    wal_file: File,
+struct Mmap {
+    ptr: *mut u8,
+    sz: usize,
 }
 
-impl WalWriter {
-    /// Open a WAL file if it exists, otherwise, create a new empty one.
-    ///
-    /// No check with be performed if the file exists,
-    /// it will only happend new batch to the end of the file
-    /// and purge from the front
-    pub fn open(wal_file: &File) -> anyhow::Result<Self> {
-        // TODO: remove this clone.
-        // Panic here for visibility.
-        let wal_file = wal_file.try_clone().unwrap();
-        Ok(Self { wal_file })
-    }
-
-    /// Get the current length of the file.
-    pub fn file_size(&self) -> u64 {
-        self.wal_file.metadata().unwrap().len()
-    }
-
-    /// Clean up the first n bytes of the file.
-    pub fn prune_front(&self, n: u64) {
-        // fallocate call with flag FALLOC_FL_COLLAPSE_RANGE
-        // returns an error if you're trying to collapse 0 bytes
-        if n == 0 {
-            return;
-        }
-
-        // TODO: other UNIXes are going to need not a WAL but a WAL
-        // file pool. FALLOC_FL_COLLAPSE_RANGE only works on Linux.
-        #[cfg(target_os = "linux")]
-        unsafe {
-            let res = libc::fallocate(
-                self.wal_file.as_raw_fd(),
-                libc::FALLOC_FL_COLLAPSE_RANGE,
+impl Mmap {
+    fn new(sz: usize) -> Self {
+        let ptr = unsafe {
+            libc::mmap(
```

Comment:
this seems correct as a way to allocate the memory needed for the WAL, but I wonder if mmap is overkill here

Reply:
I guess as it stands right now — yes. This was written for the single blob case but now we still split the whole thing into pages in the writeout.

that being said, i still hope we come back to the single blob approach and changing this now would be extra effort.

Reply:
yes, that definitely makes sense. We should discuss how to return to a single blob approach. I don't view that as a blocker for this PR.